datafusion/physical-plan/src/joins/ie_join.rs

```diff
@@ -0,0 +1,1010 @@
+            _join_type: self.join_type,
+            operators: self.operators.clone(),
+            sort_options: self.sort_options.clone(),
+            iejoin_data,
+            column_indices: self.column_indices.clone(),
+            pairs: Arc::clone(&self.pairs),
+            finished: false,
+        }))
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metrics.clone_inner())
+    }
+
+    fn statistics(&self) -> Result<Statistics> {
+        estimate_join_statistics(
+            Arc::clone(&self.left),
+            Arc::clone(&self.right),
+            vec![],
+            &self.join_type,
+            &self.schema,
+        )
+    }
+}
+
+#[derive(Debug, Clone)]
+/// SortedBlock contains arrays that are sorted by specified columns
+// TODO: use struct support spill?
+pub struct SortedBlock {
+    pub array: Vec<ArrayRef>,
+    pub sort_options: Vec<(usize, SortOptions)>,
+}
+
+impl SortedBlock {
+    pub fn new(array: Vec<ArrayRef>, sort_options: Vec<(usize, SortOptions)>) -> Self {
+        Self {
+            array,
+            sort_options,
+        }
+    }
+
+    /// sort the block by the specified columns
+    pub fn sort_by_columns(&mut self) -> Result<()> {
+        let sort_columns = self
+            .sort_options
+            .iter()
+            .map(|(i, opt)| SortColumn {
+                values: self.array[*i].clone(),
+                options: Some(*opt),
+            })
+            .collect::<Vec<_>>();
+        // TODO: should handle list type?
+        let indices = lexsort_to_indices(&sort_columns, None)?;
+        self.array = self
+            .array
+            .iter()
+            .map(|array| take(array, &indices, None))
+            .collect::<Result<_, _>>()?;
+        Ok(())
+    }
+
+    pub fn arrays(&self) -> &[ArrayRef] {
+        &self.array
+    }
+
+    pub fn slice(&self, range: Range<usize>) -> Self {
+        let array = self
+            .array
+            .iter()
+            .map(|array| array.slice(range.start, range.end - range.start))
+            .collect();
+        SortedBlock::new(array, self.sort_options.clone())
+    }
+}
+
+#[derive(Debug)]
+pub struct IEJoinData {
+    /// collected left data after sort by condition 1
+    pub left_data: Vec<RecordBatch>,
+    /// collected right data after sort by condition 1
+    pub right_data: Vec<RecordBatch>,
+    /// sorted blocks of left data, contains the evaluated result of condition 1 and condition 2
+    pub left_blocks: Vec<SortedBlock>,
+    /// sorted blocks of right data, contains the evaluated result of condition 1 and condition 2
+    pub right_blocks: Vec<SortedBlock>,
+}
+
+async fn collect_by_condition(
+    input: Arc<dyn ExecutionPlan>,
+    sort_expr: PhysicalSortExpr,
+    context: Arc<TaskContext>,
+) -> Result<Vec<RecordBatch>> {
+    let sort_plan = Arc::new(SortExec::new(vec![sort_expr], input));
+    let record_batches = collect(sort_plan, context).await?;
```

Comment:
I found the `collect` function don't collect all data from `SortExec`(it only return one recordbatch, i don't know why), which function should be use here?

Reply:
This way don't works. I add SortExec directly in planner(https://github.com/apache/datafusion/blob/4d488103287690ac4b6e0306f6e3bc84cdbfb0e2/datafusion/core/src/physical_planner.rs#L1756).